src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -502,21 +505,60 @@ static bool useHeavyMonitors() {
 
 // -----------------------------------------------------------------------------
 // Monitor Enter/Exit
+
+void ObjectSynchronizer::enter_for(Handle obj, BasicLock* lock, JavaThread* locking_thread) {
+  // When called with locking_thread != Thread::current() some mechanism must synchronize
+  // the locking_thread with respect to the current thread. Currently only used when
+  // deoptimizing and re-locking locks. See Deoptimization::relock_objects
+  assert(locking_thread == Thread::current() || locking_thread->is_obj_deopt_suspend(), "must be");
+  if (!enter_fast_impl(obj, lock, locking_thread)) {
+    // Inflated ObjectMonitor::enter_for is required
+
+    // An async deflation can race after the inflate_for() call and before
+    // enter_for() can make the ObjectMonitor busy. enter_for() returns false
+    // if we have lost the race to async deflation and we simply try again.
+    while (true) {
+      ObjectMonitor* monitor = inflate_for(locking_thread, obj(), inflate_cause_monitor_enter);
+      if (monitor->enter_for(locking_thread)) {
+        return;
+      }
+      assert(monitor->is_being_async_deflated(), "must be");
```

Comment:
You've added this assertion here after the `enter_for` must have lost the
race to async deflation, but you didn't add the same assertion below
after L540 -> L542. Why not?

Reply:
The assert was added because `ObjectMonitor::enter_for` was initially not constrained by only racing with deflation. Now however `ObjectMonitor::enter_for` does assert that it succeeds if it manage to lock out deflation. This assert should never be reached but I kept it as it shows intent.

I did not added it to the `ObjectSynchronizer::enter` at the time because it did not seem needed. But given that I kept it in 'ObjectSynchronizer::enter_for' to show intent, the same argument could be made for `ObjectSynchronizer::enter`.